RFD 0205: Improved On-Prem Joining for Machine ID

timothyb89 opened the pull request:
This RFD discusses improvements to on-prem and non-delegated bot joining, focusing on a new `bound-keypair` join method.

[Rendered](https://github.com/gravitational/teleport/blob/rfd/0205-improved-onprem-joining/rfd/0205-improved-onprem-joining.md)

timothyb89 committed:
RFD 0205: Improved On-Prem Joining for Machine ID

This RFD discusses improvements to on-prem and non-delegated bot
joining, focusing on a new `challenge` join method.

timothyb89 committed:
Various whitespace fixes

timothyb89 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,418 @@
+authentication ceremony, clients can use `go-jose` to marshal and sign a JWT
+which can then be verified easily on the server.
+
+TODO: This needs significant further elaboration and feedback.
```
Intentionally brief at the moment, I think I'd like to consult with some experts before committing to any crypto implementation. Very much open to ideas and feedback on this!

strideynet replied:
I think JWS/JWT seems to be a sensible choice here. 

How will this flow work when the bot is providing its public key on initial join? Will it still need to perform an initial signature?

timothyb89 replied:
I think so, it seems like it'd be tidier if the secret-for-pubkey exchange was effectively an independent step and then it did a full regular rejoin process.

That said, keeping them fully separate may create unnecessary complexity elsewhere, since we'd need to have yet another public RPC. I'll try to explore both options and see which is simpler and easier to keep secure.

strideynet commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,418 @@
+
+status:
+  challenge:
+    # If `public_key` is unset, this value will be generated server-side and
```
```suggestion
    # If `spec.onboarding.public_key` is unset and no value has been provided in `spec.onboarding.initial_join_secret`, this value will be generated server-side and
```

strideynet commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,418 @@
+
+#### Token Resource Example
+
+`challenge`-type tokens differ from other types in that they are intended to
```
We'll definitely want to discourage folks explicitly in documentation from setting a resource level expiry if it may break the rejoin mechanism.

timothyb89 replied:
I could see it still being useful: if I want a bot that only lasts for e.g. 1 month, it would be nice if it could expire automatically, though we'd need some additional logic during regular cert renewals to ensure the backing token still exists.

strideynet replied:
> though we'd need some additional logic during regular cert renewals to ensure the backing token still exists.

Yeah I think so - either way we just need to make sure the behaviour is fairly sane/well documented. I think worst case scenario would be someone deleting a known-bad token thinking it'll disconnect the bot - and it still has access.

strideynet commented on rfd/0205-improved-onprem-joining.md:
````diff
@@ -0,0 +1,418 @@
+$ tbot start --token=challenge:04f0ceff1bd0589ba45c1832dfc8feaf ...
+```
+
+## Alternatives and Future Extensions
````
One thing that I think would be useful to consider within the RFD is whether this process should be tied to a join token or directly to a Bot Instance. I think join token is likely the correct choice but I think worth at least considering.

timothyb89 replied:
I added a section under alternatives to discuss this. I think the most straightforward argument in favor of keeping state out of bot instances is that it'd make it harder to extend this join method to regular nodes if we ever wanted to do that.

strideynet commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,418 @@
+- Effort required to configure IaC / Terraform is still fairly high, even if
+  reduced.
+
+### Other Supporting Improvements
```
For organizations that desire multi-use tokens because it alleviates pain for a central administrator who must vend these, I think one supporting improvement would be to allow more tightly scoped rbac on token CRUD based on the role/bot_name.

timothyb89 replied:
Hmm, this would be fantastic, I'll add a section for this.

timothyb89 replied:
I've added a section, but I think this probably depends on [Scoped RBAC](https://github.com/gravitational/teleport/pull/38078)

strideynet commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,418 @@
+
+### Challenge-Response Joining
+
+TODO: Consider alternative join method names?
```
`bound-keypair`?

boxofrad replied:
Mentioning a key or keypair feels more intuitive to me because it's focussed on what the user has and sees rather than the technical implementation (the same is true of `token` and `tpm`), and will feel familiar if you've used SSH keys for authenticating with GitHub, etc.

timothyb89 committed:
Add details after first feedback pass

Adds sections on alerting, keypair rotation, and intention to
eventually support node joining.

thedevelopnik commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,454 @@
+significantly more flexibility than today's `token` join method. This works by -
+in a sense - inverting the token joining procedure: bots generate an ED25519
+keypair, and the public key is copied to the server. The public key can be
+copied out-of-band, or bots can provide their public key on first join using a
```
I feel like I'm not understanding this correctly. If they can offer their public key on first join, what's to stop a malicious actor from just joining a bunch of bots? The security of the token is that the server generates it and knows it's valid. I get adding the public key out-of-band, but not this.

Ok, reading again, in this scenario where the public key is not added ahead of time, there would be an initial join secret, so basically a token would still be needed, but after that initial contact it wouldn't need the token again?

timothyb89 replied:
For bulk joining, you'd want to use the Terraform provider with scenario 2, and have it generate an initial join secret for each node - there's an example of this further down in the document. The generated secret value can be provided to `tbot` and it will behave similarly to token joining today; keypair generation and exchange with Auth will be entirely transparent to the user.

Overall, in its default mode (i.e. scenario 2) it's functionally identical to traditional `token` joining for end users doing deployment, with a few enhancements to make Terraform automation a bit nicer (server side secret fulfillment, a sane story for restoring broken bots, etc).

timothyb89 committed:
Add section detailing joining flows, various other details

timothyb89 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,541 @@
+      # token, and creation or update will fail if a public key is reused.
+      public_key: null
+
+      # If set, use an explicit initial joining secret; if both this and
+      # `public_key` are unset, a value will be generated server-side and
+      # stored in `.status.challenge.initial_join_secret`
+      initial_join_secret: ""
+
```
Definitely, no, it's more providing for compatibility with today's workflows as `token`-type tokens effectively work this way. If we think it's problematic I have no real issue removing it.

From a convenience PoV - Terraform will happily fetch field values mutated on the server during creation, so no problems making use of server-generated secrets. But if I'm trying to automate creation with `tctl` and a bash script, it's a bit harder:

```
$ tctl create -f - <<EOF
kind: token
version: v2
metadata:
  name: my-join-token
spec:
  bot_name: example
  join_method: challenge
EOF
$ token=$(tctl get token/my-join-token --format=json | jq '.status.challenge.initial_join_secret')
$ tbot start --token=$token ...
```

It may be slightly easier to generate the secret client-side:
```
$ token=$(pwgen -s 32 1)
$ tctl create -f - <<EOF
kind: token
version: v2
metadata:
  name: my-join-token
spec:
  bot_name: example
  join_method: challenge
  onboarding:
    initial_join_secret: "$token"
EOF
$ tbot start --token=$token ...
```

timothyb89 committed:
Fix cspell nits

zmb3 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,541 @@
+  automatically share their public key with the server.
+
+- When joining or rejoining, Teleport issues a challenge that the client must
+  solve. This is similar to TPM joining today, but backed by a local keypair
+  rather than (necessarily) a hardware token.
```
_Could_ it also work with a TPM though?

timothyb89 replied:
There's a short section on this, but if it supports PKCS#11 key storage then yes, we'd like to support that by making use of our existing HSM library support. I intend to keep the storage backend extensible to make supporting this straightforward. I could also see a macOS-specific backend, for example.

timothyb89 replied:
I've added more details on this and explicitly called out support for TPMs/secure enclave.

zmb3 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,541 @@
+
+      # Initial joining must take place before this timestamp. May be
+      # modified if bot has not yet joined.
+      expires: "2025-03-01T21:45:40.104524Z"
```
Should we just use the token resource's expiry for this? Will it be confusing to have an expiry in both the metadata and the spec?

timothyb89 replied:
It's separate because we don't want the resource to get purged after the expiration - the idea is that we can re-up a bot on the server by increasing its expiration/renewal count even after it's expired... which means the token resource itself needs to remain available.

Unfortunately that means if we want some time-based expiry layer that _doesn't_ lead to the resource getting garbage collected automatically, I think it needs to be in a separate field.

timothyb89 replied:
I've renamed the two duplicate fields to `must_join_before` and `must_rejoin_before`, to hopefully reduce confusion.

strideynet replied:
Bit of a nitpick: What about `can_rejoin_before` ? or `can_rejoin_until`? "Must" makes it sound like the Bot must rejoin.

timothyb89 replied:
I've renamed the field to `may_join_until` (and standardized naming around "join" instead of "rejoin")

zmb3 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,541 @@
+
+      # If set, rejoining is only valid before this timestamp; may be
+      # incremented to extend bot lifespan.
+      expires: ""
```
Yeah, I'm thinking 3 different expiration times in the same resource is bound to get confusing.

timothyb89 replied:
Yeah... this is the biggest sore spot I think. I think I might remove the resource-level expiration (or, well, make it an error to set one), and might drop this secondary expiration. Playing with some ideas to make the ergonomics here make sense, but it's confusing as is.

timothyb89 replied:
As noted in another thread, I've renamed the duplicate fields. I'm not sure if that's ultimately a perfect solution but it's hopefully at least less confusing.

I do think (the field now known as) `must_rejoin_before` is optional. I think it's a reasonably sensible additional rejoining control admins might like to have available, but I think ultimately the design works alright with just a counter. If you think the end result is still confusing, we can remove this field.

zmb3 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,541 @@
+lengthened. We propose raising this limit to 7 days, but keeping the default
+(1hr) the same.
+
+#### Bot Instance Locking
```
Do we also want the ability to put a lock on the public key / join token to prevent compromised bots from getting new credentials?

timothyb89 replied:
Yes, it will (at least) be indirectly linked to a join token via the bot instance, but locking on public key could work too. I'll try to evaluate which avenue will work best, or maybe all if it's relatively simple to add the additional lock targets.

timothyb89 replied:
I've added some details on specific new lock targets, including public key.

timothyb89 committed:
Rename to bound keypair, address review feedback

This renames the join method to `bound-keypair`, adds sections on
extensible keystore backends, non-Terraform UX, and scoped RBAC.

timothyb89 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,541 @@
+Primarily, we discuss a new `challenge` join method intended to replace the
+traditional `token` join method for many use cases, but also proposes a number
+of UX improvements to improve bot joining generally and `token` or
+`challenge-response` joining in particular.
```
I've updated everything to refer to the new join method as `bound-keypair`.

timothyb89 committed:
Rewrite join UX improvement to use URIs

Adds a new URI joining proposal

timothyb89 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,418 @@
+
+#### Client-Side Changes in `tbot`
+
+Bots should be informed of their number of remaining rejoins. We can give bots
```
Hmm, I see no reason we couldn't extend `SubmitHeartbeatResponse` to include this, the handler would just need to look up the token, which is easy enough. I'll add it as the recommended option.

timothyb89 committed:
Rewrite some sections, discuss state storage

New sections on state storage and rejected alternatives, plus
rewrote several sections for clarity.

timothyb89 committed:
Rename overlapping `expires` fields

thedevelopnik commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -316,7 +317,9 @@ status:
-    # decremented, this value cannot fall below zero.
+    # decremented, this value cannot fall below zero. If
+    # `.spec.bound_keypair.rejoining.unlimited` is set, this value will always
+    # be 0 but rejoin attempts will be succeed.
```
```suggestion
    # be 0 but rejoin attempts will succeed.
```

FireDrunk commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,716 @@
+   precedent here as bots can view e.g. their own roles without explicitly
+   having RBAC permissions to do so.
+
+The remaining rejoin counter should then be exposed as a Prometheus metric to
```
Perhaps the last time it (re)joined would be a good addition to the metrics as well.

timothyb89 replied:
It's a bit tricky to publish date fields directly in prometheus, but I think we could publish a metric like this:
```
teleport_bot_bound_keypair_joins{bot_name="example"} 1
```

You could then alert against it with a query like this:
```
sum by (user) (teleport_bot_bound_keypair_joins[1h]) > 1
```

Prometheus should count each increase in the field and will account for counter resets if auth restarts. I think this would meet your needs?

FireDrunk replied:
Perhaps a good middle-ground would be `teleport_bot_rejoined_seconds` where it has an ever increasing value with the number of seconds since the last rejoin.
Creating an alert would be as simple as checking if the value is less than 1h.

I think a metric like `system_uptime` would be comparable.

FireDrunk commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,716 @@
+
+We should take steps to improve visibility of bots at or near expiry, including:
+
+- Configurable cluster alerts when the number of available renewals has crossed
```
Maybe add an alert that states something like: "A bot has recently automatically rejoined", without specifically watching the count.
Where 'recently' is configurable.

timothyb89 replied:
That's a good thought, I've added a note about it to this section. There's some mild complexity here, I think, since alerts could get noisy if there's hundreds (or more) bots active. I think we'll definitely need to iterate on this as we see how this method gets used.

FireDrunk commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,716 @@
+  token. This may require introduction of a new certificate field to track the
+  exact join token used.
+
+- Public key locking: locks bots joining with a particular public key. A
```
Another option would be to allow a bot to be limited to only a single instance. If a second instance joins, the first one is revoked (by design).

Another option would be to only allow rejoining if all instances are rejoining at once. Say after a power outage on a site with multiple instances. The logic would validate that there are other instances in the same bot, and only allow a generation increase if they all 'agree'.

timothyb89 replied:
I've tried to further clarify the approach here in the "bot lifecycle" section since it was vague and wasn't fully written down, and actually pulling on this thread led to introducing "Join State Documents" in the new "Preventing Credential Duplication" section.

The short version is, we'll only allow one active bot instance per `(bot, bound-keypair token)` pair. If a rejoin occurs, the previous instance will not be allowed to refresh its certificates any further and will need to rejoin - and might additionally get locked out. If a bot's keypair is cloned and 2 clients start competing to rejoin, we'll detect this and lock all bots using the join token, similar to generation counter lockouts today.

FireDrunk commented on rfd/0205-improved-onprem-joining.md:
````diff
@@ -0,0 +1,716 @@
+
+The URI syntax might look like this:
+```
+tbot+[auth|proxy]://[join method]:[token value]@[addr]:[port]?key=val&foo=bar
````
For easy of copying this, you could base64 encode this and/or even sign it like a JWT.

timothyb89 replied:
I think there are some benefits to it being cleartext, mainly ease of confirming bots are routed to the right place, and ability to tweak the URI if needed, e.g. connecting to a leaf cluster. From a UX perspective, it's still functionally a single token for copy and paste purposes, which Teleport can fully compose it for you in the CLI/web. Do you see a use case for shorter connection strings? 

I'm not too sure of the value of signing these - could you elaborate on the use case a bit? Bots still verify TLS on startup, and in general I don't see bots connecting to a hostile Teleport instance as a meaningful threat vector.

FireDrunk replied:
A token is easily copy-pastable in a shell (i.e. vim/bash etc), where a complex url usually is a bit more annoying. Especially if you have `+` and `&` characters.
Having those replaced via base64 encoding significantly reduces the amount of user error in most cases.

Signing would just be to validate that you have copied the whole string. Maybe I should have called it hashing.

FireDrunk commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,716 @@
+joining as well as bots, as a more secure alternative to static or long-lived
+join tokens.
+
+### Additional Keypair Protections
```
Would you allow CA based joining? In that case someone could pregenerate the tbot private keypair and sign it from a centralized CA.
One could then add that public CA to the ProvisionToken as allowed to rejoin.

That way all tbot's could rejoin if they have a certificate signed by a single matching CA.

FireDrunk commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,716 @@
+URL paths and query parameters may also provide options for future extension if
+desired.
+
+## Future Extensions and Alternatives
```
Would you consider a little scope creep and add extra rejoin 'conditions'?
I would for instance think that rejoining on the same IP address could automatically be allowed. It's not much, but rejoining from a different IP should (in our use-case) automatically be denied.

The 'extension' could be:
- Allow external factors or 'proof' to be sent along with a rejoin request to automatically allow it.

Things I was thinking about:
- (Soft)TPM joining (if you do not want _auto_ TPM join, just a limited amount, and still want to do custom keypairs)
- Public IP Address
- Physical Keys
- Information it can exfiltrate from the place it's running. For instance a challenge against a networked HSM (Hashicorp Vault?).

timothyb89 replied:
I think there's definitely some room to add additional joining requirements over time, public IP in particular could be helpful.

I think your "soft TPM" idea might be covered by the proposal already, if I understand correctly? That sounds like the TPM/HSM private key storage backend?

timothyb89 replied:
I think this is an interesting flow for sure, though I do think we might have something mostly equivalent with the initial join secret + HSM storage:

1. Create bound keypair token with initial join secret
2. Provision disk with initial join secret
3. On boot, the machine generates a keypair on the TPM/HSM and establishes trust with Auth using the join secret
4. For each join, auth issues a challenge that must be completed using the keypair on the TPM/HSM 
5. If desired, the keypair can be rotated at any time, which would create a new HSM-stored keypair and switch Auth's trust to it.

My understanding is that a HSM or TPM-stored keypair should be more or less equivalent to the module's built-in key.

I think the main limitation here is that through this method Auth can't 100% trust that the client is in fact using a hardware keystore, but I'd argue that's strictly true for any automatic TPM enrollment.

It's probably out of scope for this first revision, but it might be interesting to add some additional challenge requirements in the future, like an additional EKCert attestation or similar. I think that's the only way we could (kind of) ensure a real TPM was in use, but that'd need some more in depth design.

timothyb89 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,716 @@
+      # skipped and instance will directly prove its identity using its private
+      # key. It is an error for a public key to be associated with more than one
+      # token, and creation or update will fail if a public key is reused.
+      public_key: null
```
self nit, rename this to `initial_public_key`

timothyb89 committed:
Pivot to delegated joining impl. Add sequence diagram.

timothyb89 commented on rfd/0205-improved-onprem-joining.md:
Somewhat significant update after a discussion with @strideynet today - we're tentatively pivoting toward implementing this as a delegated join method, meaning bots will complete the challenge ceremony on every renewal attempt.

This means we won't use traditional renewable certs at all, and we determine if a particular join attempt costs a rejoin credit (probably need a better term for this) based on whether or not the bot presents an existing client certificate. We already use this mechanism to preserve bot instance IDs today.

This should keep the implementation more in line with other delegated joining methods (i.e. all of them except token) and could put us on a path toward eliminating traditional renewable certs in the future.

timothyb89 replied:
Quick summary of additional updates today:
* Added credential duplication mitigation (aka generation counter lite)
* Added keypair rotation procedure
* Added bot lifecycle details
* Tweaked the joining URL scheme to allow for both token name + additional parameter (i.e. initial joining secret)

I think my only outstanding item is that I'm thinking of renaming "rejoining" to just "joining", especially in the token spec. It's really the same process and I think it's conceptually simpler if we don't special case onboarding - that's a join, just like an expired connection.

timothyb89 replied:
Another small summary of updates:
- Renamed `rejoining` to `joining` in most contexts, since it's the same process.
- Expanded join state document, added example. During a discussion with @boxofrad and @strideynet, we decided to move the remaining join counter and rotation flag into the document as well.
- Added notes about locking old instances after a rejoin. There's probably some value here, but I think it could cause disruptions in apps as well.

@strideynet also suggested exclusively using the join state sequence number for all joins and refreshes, instead of using it for joins, and then using the standard generation counter (stored in the bot instance) for refreshes. I think there's some merit to this and haven't come to a firm conclusion yet. We agreed it's okay as written, but I may change it here after thinking on it further.

timothyb89 committed:
Don't assume ED25519; fix renew->refresh terminology

strideynet reviewed (commented):
I think I'm at a point where I'm happy with the overall design of this now - we probably want to begin implementation before merging down this RFD to see if anything shakes out.

timothyb89 committed:
Tweak joining URL scheme

Moves join method to the URL scheme to allow joining secrets; more
examples added.

timothyb89 committed:
Bot lifecycle; removed uniqueness requirement

Removes "soft bot expiration" section as this has been resolved
with the switch to delegated joining. Also added a Bot Lifecycle
section to describe how bots are expected to be disabled.

Also removed the public key uniqueness requirement. At join time bots
now specify both the token name and joining secret (if any), so we
won't need to search all tokens for a matching key. It was also not
efficient to ensure uniqueness among all provision tokens.

timothyb89 committed:
Add keypair rotation details

timothyb89 committed:
Credential duplication mitigation, proto draft

Describes a method for mitigating credential duplication, and
includes a protobuf draft.

boxofrad commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,1010 @@
+      # expired.
+      may_rejoin_until: "2026-03-01T21:45:40.104524Z"
+
+    # If set, the bot will perform a keypair rotation on its next renewal after
+    # it is informed of the change to this field. Note that this is tied to bot
+    # heartbeats and may not take effect on the next refresh interval.
+    rotate_on_next_renewal: false
```
Does this field get automatically reset when the bot rotates its public key?

I'm wondering if it's possible for an admin to set this and forget about it (leading to rotations on every renewal). Which arguably wouldn't be the worst thing in the world, just a little wasteful.

timothyb89 replied:
Writing down what we discussed, yes, this field should be cleared after a rotation. It's alluded to in the flowchart, but I'll update the doc to make that clear.

boxofrad commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,1010 @@
+service JoinService {
+  // ...snip...
+
+  rpc RegisterUsingBoundKeypair(stream RegisterUsingBoundKeypairRequest) returns (stream RegisterUsingBoundKeypairResponse);
```
nit: What's the benefit of this being a streaming RPC rather than a series of unary RPCs?

timothyb89 replied:
Just to write down some reasoning from the verbal discussion:
- Most other methods with a challenge-response process (e.g. TPM joining) do this already
- Consistently ties a renewal ceremony to one Auth instance if behind a LB, which might be important for consistency on some backends
- Is mildly nicer to keep things within a single socket session when multiple (potentially optional) steps may be involved in a join attempt, e.g. keypair rotation.

boxofrad commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,1010 @@
+The remaining rejoin counter should then be exposed as a Prometheus metric to
+allow for alerting if a bot drops below some threshold.
+
+Importantly, this is a potentially lagging indicator. The design allows for the
+rejoin counter to be decreased (to zero) at any time, so a rejoin attempt may
+still fail at any time. This should be acceptable since it can also be increased
+after the fact to restore access if desired.
```
Given that we need to handle this failure-mode anyway, is there much benefit to pushing the rejoin counter and whether a key rotation is required to the bot proactively? (as in the heartbeating suggestion above).

Instead, we could just return the rejoin counter on each renewal, and return a "you need to rotate your public key" response when trying to renew after `rotate_on_next_renewal` has been set.

timothyb89 replied:
Following up on our discussion, I think the rejoin counter could be included in the join state document, and I'll at least move the `rotate_on_next_renewal` flag here. I think the forced-rotation-via-error flow has some merits but for simplicity of implementation I have a mild preference to let clients initiate the rotation process.

boxofrad reviewed (approved):
This is looking great, @timothyb89. Amazing work! Looking forward to chatting it through later today.

timothyb89 committed:
Rename most references to "rejoining" to just "joining"

These were fundamentally the same processes, so we'll standardize on
calling both initial joining and rejoining different modes of just
"joining".

timothyb89 committed:
Add example join state document

timothyb89 committed:
Add note about locking old instance after rejoin

timothyb89 committed:
Small fixes

timothyb89 committed:
Fix word missing from cspell

hugoShaka commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,1060 @@
+  expect your system to never go down for more than 24 hours, bots can happily
+  run for months.
+
+  (Ironically, Kubernetes is a great environment in which to run `token`-joined
+  bots since it'll rapidly reschedule any bot deployments that fail... but we
+  have a dedicated `kubernetes` delegated join method.)
```
I disagree, Kubernetes is not a viable environment for token joining. This misconception caused us to build the kube secret state persistence and do irremediable harm to the teleport-kube-agent by making it stateful.

I think this claim can be removed without altering the RFD's essence. I don't want to bikeshed on topics which are not central to the RFD, feel free to Slack me if you want to exchange about token joining challenges in kube.

timothyb89 replied:
Heh, that's a fair call out, I've removed that comment.

That said, I'd definitely like to hear more about the issues you ran into. Statelessness is definitely the ideal, especially for your use case, but I'd like this to be as viable as possible. The hope is that this new join method would have made your experience at least _less_ miserable.

timothyb89 committed:
Remove kubernetes reference

timothyb89 committed:
Fix hanging sentence, add reference to join state document in lifecycle

timothyb89 committed:
SSH key format, introduce "insecure" flag

timothyb89 commented on rfd/0205-improved-onprem-joining.md:
```diff
@@ -0,0 +1,1081 @@
+      # expired.
+      may_join_until: "2026-03-01T21:45:40.104524Z"
+
+      # Insecure is an optional flag that enables insecure joining and
+      # rejoining. This method disables generation counter checks during joining
+      # and rejoining. When combined with the `unlimited` flag, this allows
+      # unlimited reuse of this token provided the client has access to the
+      # keypair. This may be useful in certain cases - like use in unsupported
+      # CI/CD providers - but cannot offer the same security assurances and
+      # should be used with care.
+      insecure: false
```
Small update: this is something we've mulled over for a while and was loosely mentioned in a section below ("explicitly insecure token joining"). I think we can accomplish a lightweight version of that with just this additional flag; users would enable both `unlimited` and `insecure`, and could then share a keypair in e.g. a CI/CD secret.

We've had a number of customers - especially those using unsupported CI/CD platforms - looking to take some advantage of Machine ID. I've added it now since I think now (while I'm WIP on the initial implementation of this RFD) is the most efficient time to add the flag, but am happy to cut this if there's any concern.

viktor-doyensec-teleport commented:
Hello team! 👋 

I've just completed my review of the RFD.

The RDF explains in detail all the proposed changes to the system, while taking potential abuse cases into consideration, and implementing sufficient security controls. I believe that, given proper implementation, no notable vulnerabilities will be introduces.

Overall, the RDF looks good to me!

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into rfd/0205-improved-onprem-joining

timothyb89 committed:
Mark as implemented

timothyb89 committed:
Various updates to reflect implemented state

Of note:
- s/rejoin/recovery (give or take some conjugation)
- Updated keypair rotation and sequence diagram
- Updated join state spec
- Updated token resource example to match public docs
- Lots of misc terminology tweaks
